silkrpc/ethdb/kv/state_cache.cpp

```diff
@@ -0,0 +1,368 @@
+    SILKRPC_DEBUG << "Miss in code cache: lookup in Code key=" << key << " value=" << value << "\n";
+    if (value.empty()) {
+        co_return std::nullopt;
+    }
+
+    read_lock.unlock();
+    std::unique_lock write_lock{rw_mutex_};
+
+    add_code({key, value}, root_it->second.get(), view_id);
+
+    co_return value;
+}
+
+CoherentStateRoot* CoherentStateCache::get_root(StateViewId view_id) {
+    const auto root_it = state_view_roots_.find(view_id);
+    if (root_it != state_view_roots_.end()) {
+        SILKRPC_DEBUG << "CoherentStateCache::get_root view_id=" << view_id << " root=" << root_it->second.get() << " found\n";
+        return root_it->second.get();
+    }
+    const auto [new_root_it, _] = state_view_roots_.emplace(view_id, std::make_unique<CoherentStateRoot>());
+    SILKRPC_DEBUG << "CoherentStateCache::get_root view_id=" << view_id << " root=" << root_it->second.get() << " created\n";
+    return new_root_it->second.get();
+}
+
+CoherentStateRoot* CoherentStateCache::advance_root(StateViewId view_id) {
+    CoherentStateRoot* root = get_root(view_id);
+
+    const auto previous_root_it = state_view_roots_.find(view_id - 1);
+    if (previous_root_it != state_view_roots_.end() && previous_root_it->second->canonical) {
+        SILKRPC_DEBUG << "CoherentStateCache::advance_root canonical view_id-1=" << (view_id - 1) << " found\n";
+        root->cache = previous_root_it->second->cache;
+        root->code_cache = previous_root_it->second->code_cache;
+    } else {
+        SILKRPC_DEBUG << "CoherentStateCache::advance_root canonical view_id-1=" << (view_id - 1) << " not found\n";
+        state_evictions_.clear();
+        for (const auto& kv : root->cache) {
+            state_evictions_.push_front(kv);
+        }
+        code_evictions_.clear();
+        for (const auto& kv : root->code_cache) {
+            code_evictions_.push_front(kv);
+        }
+    }
+    root->canonical = true;
+
+    evict_roots(view_id);
+
+    latest_state_view_id_ = view_id;
+    latest_state_view_ = root;
+
+    state_eviction_count_ = state_evictions_.size();
+    code_eviction_count_ = code_evictions_.size();
+
+    return root;
+}
+
+void CoherentStateCache::evict_roots(StateViewId next_view_id) {
+    SILKRPC_DEBUG << "CoherentStateCache::evict_roots state_view_roots_.size()=" << state_view_roots_.size() << "\n";
+    if (state_view_roots_.size() <= config_.max_views) {
+        return;
+    }
+    if (next_view_id == 0) {
+        // Next view ID is zero with cache not empty => view ID wrapping => clear the cache except for new latest view
+        std::erase_if(state_view_roots_, [&](const auto& item) {
+            auto const& [view_id, _] = item;
+            return view_id != next_view_id;
+        });
+        return;
+    }
+    // Erase older state views in order not to exceed max_views
+    const auto max_view_id_to_delete = latest_state_view_id_ - config_.max_views + 1;
```

Comment:
if we use: max_view_id_to_delete = latest_state_view_id_ - config_.max_views + 1;
the view_id < max_view_id_to_delete;
or 
if we use: max_view_id_to_delete = latest_state_view_id_ - config_.max_views;
the view_id <= max_view_id_to_delete;

Reply:
This is not the case because `evict_roots` is called before `latest_view_id_` gets updated